fix: renamed numpy_batch to numpy_sample throughout the codebase

Tim1119 opened the pull request:
# Pull Request

## Description
This pull request renames the `numpy_batch` folder to `numpy_sample` because we only deal with samples, not batches. All relevant references and imports have been updated to reflect this change. The modification aligns the naming convention with the actual functionality for better clarity and consistency.

Fixes #103

## How Has This Been Tested?
- Ran unit tests to verify functionality after renaming the directory.
- Checked all imports and ensured there were no unresolved references.
- Verified that the application runs without errors after the changes.

- [x] Yes

The renaming does not change the logic of data processing, but it does ensure the terminology accurately reflects the operations being performed. I have verified that it runs successfully after the changes, ensuring there is no disruption in functionality.

- [x] Yes


## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have checked my code and corrected any misspellings.

peterdudfield commented:
Thanks so much @Tim1119 for doing this and jumping on this so quickly

felix-e-h-p commented on ocf_data_sampler/numpy_sample/collate.py:
```diff
@@ -0,0 +1,79 @@
+logger = logging.getLogger(__name__)
+
+
+def stack_np_examples_into_sample(dict_list):
```
Suggestion for refactoring original function - just for any modification / maintainability concerns considering dealing with overall processing stages for nested data. Worth noting that core behaviour and related testing remains exactly the same.

```python

def stack_np_examples_into_sample(dict_list):
    if not dict_list:
        raise ValueError("Input empty")

   # Extract keys from first dict - structure
    sample = {}
    sample_keys = list(dict_list[0].keys())

   # Process - handle NWP separately due to nested structure
    for sample_key in sample_keys:
        if sample_key == "nwp":
            sample["nwp"] = process_nwp_data(dict_list)
        else:
            sample[sample_key] = stack_data_for_key(dict_list, sample_key)
    return sample


def process_nwp_data(dict_list):
    nwp_sample = {}
    nwp_sources = dict_list[0]["nwp"].keys()

   # Stack data for each NWP source independently
    for nwp_source in nwp_sources:
        nwp_sample[nwp_source] = stack_nested_keys(dict_list, "nwp", nwp_source)
    return nwp_sample


def stack_data_for_key(dict_list, sample_key):
    # Stack arrays for given key across all dicts
    return stack_data_list([d[sample_key] for d in dict_list], sample_key)


def stack_nested_keys(dict_list, parent_key, child_key):
    # Processes two-level nested data 
    # Preserves hierarchy whilst stacking arrays
    nested_keys = dict_list[0][parent_key][child_key].keys()
    return {
        key: stack_data_list(
            [d[parent_key][child_key][key] for d in dict_list],
            key,
        )
        for key in nested_keys
    }

```

dfulu replied:
@felix-e-h-p thanks for the suggestion. However, in my opinion this implementation is a bit overly complex. My thought are:

- `stack_nested_keys(dict_list, parent_key, child_key)` is only ever run with `parent_key="nwp"`. I think its unlikely we will use it with a different `parent_key`. I think this could just go in `process_nwp_data`
- `stack_data_for_key()` is a one liner and only used once, so maybe it doesn't need to be split out

felix-e-h-p commented on ocf_data_sampler/numpy_sample/gsp.py:
```diff
@@ -0,0 +1,33 @@
+    y_osgb = 'gsp_y_osgb'
+
+
+def convert_gsp_to_numpy_sample(da: xr.DataArray, t0_idx: int | None = None) -> dict:
```
Not strictly necessary - but perhaps validation check for required input variables? Safe measure

felix-e-h-p commented on ocf_data_sampler/numpy_sample/satellite.py:
```diff
@@ -0,0 +1,30 @@
+    t0_idx = 'satellite_t0_idx'
+
+
+def convert_satellite_to_numpy_sample(da: xr.DataArray, t0_idx: int | None = None) -> dict:
```
Likewise. Not strictly necessary - but perhaps validation check for required input variables? Safe measure

felix-e-h-p commented on ocf_data_sampler/numpy_sample/site.py:
```diff
@@ -0,0 +1,29 @@
+    id = "site_id"
+
+
+def convert_site_to_numpy_sample(da: xr.DataArray, t0_idx: int | None = None) -> dict:
```
Likewise. Not strictly necessary - but perhaps validation check for required input variables? Safe measure

felix-e-h-p commented on ocf_data_sampler/numpy_sample/nwp.py:
```diff
@@ -0,0 +1,43 @@
+    x_osgb = 'nwp_x_osgb'
+
+
+def convert_nwp_to_numpy_sample(da: xr.DataArray, t0_idx: int | None = None) -> dict:
```
Potential improvement of function example as follows - purely considering the 'TO DO' comment and related modularity there primarily. 

```python

def convert_nwp_to_numpy_sample(da: xr.DataArray, t0_idx: int | None = None) -> dict:
    """Convert from Xarray to NWP NumpySample"""
    
    # Create example and add t if available
    example = create_data(da)
    
    if "target_time_utc" in da.coords:
        example[NWPSampleKey.target_time_utc] = da.target_time_utc.values.astype(float)
    
    # Add OSGB coordinates if available and / or necessary
    add_osgb(example, da)
    
    if t0_idx is not None:
        example[NWPSampleKey.t0_idx] = t0_idx
        
    return example


def create_data(da: xr.DataArray) -> dict:
    return {
        NWPSampleKey.nwp: da.values,
        NWPSampleKey.channel_names: da.channel.values,
        NWPSampleKey.init_time_utc: da.init_time_utc.values.astype(float),
        NWPSampleKey.step: (da.step.values / pd.Timedelta("1h")).astype(int),
    }


def add_osgb(example: dict, da: xr.DataArray) -> None:
    osgb_coords = (
        (NWPSampleKey.y_osgb, "y_osgb"),
        (NWPSampleKey.x_osgb, "x_osgb"),
    )
    
    for sample_key, dataset_key in osgb_coords:
        if dataset_key in da.coords:
            example[sample_key] = da[dataset_key].values

```

dfulu replied:
I think splitting it out like this is an unnecessary change since the functions are small and only used within this one. 

If we do do this change, we should also do it in `ocf_data_sampler/numpy_sample/satellite.py:: convert_satellite_to_numpy_sample` to keep things consistent

felix-e-h-p commented on tests/numpy_sample/test_gsp.py:
```diff
@@ -0,0 +1,21 @@
+
+from ocf_data_sampler.numpy_sample import convert_gsp_to_numpy_sample, GSPSampleKey
+
+def test_convert_gsp_to_numpy_sample(uk_gsp_zarr_path):
```
Suggestion for broader testing here. Updates provide testing of all components i.e. input structure, types, capacity etc. As follows:

```python
def test_convert_gsp_to_numpy_sample(uk_gsp_zarr_path):
    
    da = (
        open_gsp(uk_gsp_zarr_path)
        .isel(time_utc=slice(0, 10))
        .sel(gsp_id=1)
    )

    numpy_sample = convert_gsp_to_numpy_sample(da)

    # Test data structure
    assert isinstance(numpy_sample, dict), "Should be dict"
    assert set(numpy_sample.keys()).issubset({
        GSPSampleKey.gsp,
        GSPSampleKey.nominal_capacity_mwp,
        GSPSampleKey.effective_capacity_mwp,
        GSPSampleKey.time_utc,
        GSPSampleKey.t0_idx
    }), "Unexpected keys"

    # Assert data content and capacity values
    assert np.array_equal(numpy_sample[GSPSampleKey.gsp], da.values), "GSP values mismatch"
    assert isinstance(numpy_sample[GSPSampleKey.time_utc], np.ndarray), "Time UTC should be numpy array"
    assert numpy_sample[GSPSampleKey.time_utc].dtype == float, "Time UTC should be float type"
    assert numpy_sample[GSPSampleKey.nominal_capacity_mwp] == da.isel(time_utc=0)["nominal_capacity_mwp"].values
    assert numpy_sample[GSPSampleKey.effective_capacity_mwp] == da.isel(time_utc=0)["effective_capacity_mwp"].values

    # Test with t0_idx
    t0_idx = 5
    numpy_sample_with_t0 = convert_gsp_to_numpy_sample(da, t0_idx=t0_idx)
    assert numpy_sample_with_t0[GSPSampleKey.t0_idx] == t0_idx, "t0_idx not correctly set"


```

Tim1119 commented:
> Thanks so much @Tim1119 for doing this and jumping on this so quickly

You're welcome

Tim1119 commented:
Thank you for the suggestions @felix-e-h-p . I'll implement the changes, refactoring the function into modular helpers as recommended while ensuring the core behavior and tests remain the same.

Tim1119 commented:
Hi @felix-e-h-p  and @peterdudfield, happy new year.  I've implemented the suggested changes and resolved the issues. Let me know if there's anything else needed or if this is ready to be merged. Thanks

peterdudfield commented:
Thanks so much for this. I know @Sukh-P and @dfulu are using this code a lot, so I wont mind them checking, if thats ok

Tim1119 commented:
> Thanks so much for this. I know @Sukh-P and @dfulu are using this code a lot, so I wont mind them checking, if thats ok

Thank you, @peterdudfield . That sounds good. I'll wait for @Sukh-P and @dfulu to review it and will address any additional feedback if needed.

dfulu commented on ocf_data_sampler/numpy_sample/collate.py:
```diff
@@ -0,0 +1,86 @@
+    }
+
+def _key_is_constant(sample_key):
+    is_constant = sample_key.endswith("t0_idx") or sample_key == NWPSampleKey.channel_names
```
I think it is better style to just return this directly rather than to assign to a variable and then return

dfulu replied:
I just realised this assignment and return was in the original function. So perhaps don't worry about this in this PR unless you want to leave things cleaner than you found them :)

dfulu commented on ocf_data_sampler/numpy_sample/gsp.py:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    # Check for None values and raise error with specific variable name if value is not passed or none
+    for var, value in input_variables.items():
```
These checks are redundant. Since the input is a `xarray.DataArray` there is no way that any of the values in `input_variables` could end up being `None`.  I think the [previous version of this function](https://github.com/openclimatefix/ocf-data-sampler/blob/main/ocf_data_sampler/numpy_batch/gsp.py#L23) which builds the example dictionary directly is sufficient and cleaner

dfulu commented on ocf_data_sampler/numpy_sample/site.py:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    # Check for None values and raise error with specific variable name if value is not passed or None
+    for var, value in input_variables.items():
```
Similar to the previous comment. These checks are redundant and I would prefer the previous version

dfulu commented on tests/numpy_sample/test_gsp.py:
```diff
@@ -0,0 +1,39 @@
+        GSPSampleKey.nominal_capacity_mwp,
+        GSPSampleKey.effective_capacity_mwp,
+        GSPSampleKey.time_utc,
+        GSPSampleKey.t0_idx
```
Strictly speaking, since the call to `convert_gsp_to_numpy_sample` did not set the `t0_idx` parameter, there should be no `GSPSampleKey.t0_idx` key. I think its better to remove this key from this set

dfulu reviewed (changes requested):
@Tim1119 thanks for all this work. There a just a couple of things which I think should be cleaned up, but besides that its all looking good!

dfulu commented:
@peterdudfield when this is merged we must remember to put a issue into PVNet since the change from XBatchKey to XSampleKey will break compatibility

Tim1119 commented:
Good day @peterdudfield ,

 I have implemented the changes suggested by @dfulu .

### Summary of Changes
1. Return value directly in ` _key_is_constant` function  for improved readability
2. Moved logic from `add_osgb` and `create_data` back to `convert_nwp_to_numpy_sample` function for simpicity and consistency
3. Removed redundant None checks in `convert_gsp_to_numpy_sample` and `convert_site_to_numpy_sample` functions
4. Removed `GSPSampleKey.t0_idx` in `test_gsp.py`  since the call to `convert_gsp_to_numpy_sample` did not set the `t0_idx parameter`
5. Moved `stack_nested_keys(dict_list, parent_key, child_key)` logic to `process_nwp_data`. Additionally I moved `stack_data_for_key()` to `stack_np_examples_into_sample` as suggested

### Testing
I ensured that all tests ran successfully before and after making the changes.

### Review Status:
Waiting for review from @Sukh-P

dfulu commented:
@Tim1119 thanks for all the additional work. I think its looking really good now. Could you have a look at the merge conflicts and then we should be ready to go 🚀

Tim1119 commented:
Hello @peterdudfield, good day, I have successfully merged all conflicts in the following files
1. `ocf_data_sampler/numpy_sample/nwp.py`
2. `ocf_data_sampler/torch_datasets/process_and_combine.py`
3. `constants.py`

Additionally, I modified all instances of "batch" to "sample" (e.g., nwpbatch to nwpsample) for consistency across the codebase, including in  newly created `tests/torch_datasets/test_process_and_combine.py.`

## How Has This Been Tested?
- Ran unit tests to verify functionality after renaming the directory.
- Checked all imports and ensured there were no unresolved references.
- Verified that the application runs without errors after the changes.

- [x] Yes

The renaming does not change the logic of data processing, but it does ensure the terminology accurately reflects the operations being performed. I have verified that it runs successfully after the changes, ensuring there is no disruption in functionality.

- [x] Yes


## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have checked my code and corrected any misspellings.

Tim1119 commented:
Hi @peterdudfield and @peterdudfield,

I noticed that several tests are failing on the GitHub Actions pipeline due to a `FileNotFoundError` for the file `non_hrv_shell.zarr.zip`, which is expected to be located at:
`file:///home/runner/work/ocf-data-sampler/ocf-data-sampler/tests/test_data/non_hrv_shell.zarr.zip`.

Since my local tests ran successfully and I didn't make any changes related to this file, I believe the issue is related to the missing test data in the CI/CD environment.

Could you please check if the file is available in the correct location in the repository or if any configuration steps might have been missed in the pipeline setup?

Thank you

peterdudfield commented:
> Hi @peterdudfield ,@Sukh-P and @dfulu
> 
> I noticed that several tests are failing on the GitHub Actions pipeline due to a `FileNotFoundError` for the file `non_hrv_shell.zarr.zip`, which is expected to be located at: `file:///home/runner/work/ocf-data-sampler/ocf-data-sampler/tests/test_data/non_hrv_shell.zarr.zip`.
> 
> Since my local tests ran successfully and I didn't make any changes related to this file, I believe the issue is related to the missing test data in the CI/CD environment.
> 
> Could you please check if the file is available in the correct location in the repository or if any configuration steps might have been missed in the pipeline setup?
> 
> Thank you

Ah i see, Ill have a go at fixing that on `main`, then you can merge in

peterdudfield commented:
Once this is merged in https://github.com/openclimatefix/ocf-data-sampler/pull/117, this will work

peterdudfield commented:
I hope its ok @Tim1119, merged this to a branch, so we could push this through
